common/src/main/java/org/apache/nemo/common/DirectByteBufferOutputStream.java

```diff
@@ -0,0 +1,164 @@
+ * Memory is allocated when needed by the specified {@code pageSize}.
+ */
+public final class DirectByteBufferOutputStream extends OutputStream {
+
+  private final LinkedList<ByteBuffer> dataList = new LinkedList<>();
+  private static int pageSize = 4096;
+  private ByteBuffer currentBuf;
+
+  /**
+   * Default constructor.
+   * Sets the {@code pageSize} as default size of 4KB.
+   */
+  public DirectByteBufferOutputStream() {
+    this(pageSize);
+  }
+
+  /**
+   * Constructor specifying the {@code size}.
+   * Sets the {@code pageSize} as {@code size}.
+   *
+   * @param size should be a power of 2 and greater than or equal to 4096.
+   */
+  public DirectByteBufferOutputStream(final int size) {
+    if (size < 4096 || (size & (size - 1)) != 0) {
+      throw new IllegalArgumentException("Invalid pageSize");
+    }
+    pageSize = size;
+  }
+
+  /**
+   * Allocates new {@link ByteBuffer} with the capacity equal to
+   * {@code pageSize}.
+   */
+  private void newLastBuffer() {
+    dataList.addLast(ByteBuffer.allocateDirect(pageSize));
+  }
+
+  /**
+   * Writes the specified byte to this output stream.
+   *
+   * @param   b   the byte to be written.
+   */
+  @Override
+  public void write(final int b) {
+    currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    if (currentBuf == null || currentBuf.remaining() <= 0) {
+      newLastBuffer();
+      currentBuf = dataList.getLast();
+    }
+    currentBuf.put((byte) b);
+  }
+
+  /**
+   * Writes {@code b.length} bytes from the specified byte array to this output stream.
+   *
+   * @param b the byte to be written.
+   */
+  @Override
+  public void write(final byte[] b) {
+    write(b, 0, b.length);
+  }
+
+  /**
+   * Writes {@code len} bytes from the specified byte array
+   * starting at offset {@code off} to this output stream.
+   *
+   * @param   b     the data.
+   * @param   off   the start offset in the data.
+   * @param   len   the number of bytes to write.
+   */
+  @Override
+  public void write(final byte[] b, final int off, final int len) {
+    int byteToWrite = len;
+    int offset = off;
+    currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
+    while (byteToWrite > 0) {
+      if (currentBuf == null || currentBuf.remaining() <= 0) {
+        newLastBuffer();
+        currentBuf = dataList.getLast();
+      }
+      final int bufRemaining = currentBuf.remaining();
+      if (bufRemaining < byteToWrite) {
+        currentBuf.put(b, offset, bufRemaining);
+        offset += bufRemaining;
+        byteToWrite -= bufRemaining;
+      } else {
+        currentBuf.put(b, offset, byteToWrite);
+        offset += byteToWrite;
+        byteToWrite = 0;
+      }
+    }
+  }
+
+  /**
+   * Creates a byte array that contains the whole content currently written in this output stream.
+   * Note that this method causes array copy which could degrade the overall performance.
+   *
```

Comment:
Could you please create an issue related to our discussion and add the jira number here? We should implement ByteBufferInputStream for performance optimization in the future and remove this method.

Reply:
That would be nice, thank you! I will update issue for this.